File: Practice/DFS_BFS/q18/q18.py
import sys
sys.setrecursionlimit(10000)

def check_correct_s(parentheses):
    stack = []
    for p in parentheses:
        if p == "(":
            stack.append(p)
        else:
            if len(stack) == 0:
                return False
            stack.pop()
    return True

def reorganize_u(u):
    reorg_u = ""
    for p in u:
        if p == "(":
            reorg_u += ")"
        else:
            reorg_u += "("
    return reorg_u

def reorganize_parentheses(parentheses):
    if parentheses == "":
        return parentheses
    l_count = 0
    r_count = 0
    s_point = None
    for idx, p in enumerate(parentheses):
        if p == "(":
            l_count += 1
        else:
            r_count += 1
        if l_count == r_count:
            s_point = idx+1
            break
    u = parentheses[:s_point]
    v = parentheses[s_point:] if s_point != len(parentheses) else ""
    if check_correct_s(u):
        return u + reorganize_parentheses(v)
    return "(" + reorganize_parentheses(v) + ")" + reorganize_u(u[1:-1])

def solution(p):
    return reorganize_parentheses(p)

File: Practice/DFS_BFS/q18/q18.cpp
#include <iostream>
#include <stack>
#include <string>

using namespace std;

bool check_correct_str(string p){
    stack<char> s;
    for(int i = 0; i < p.length(); i++){
        char c = p[i];
        if(c == '('){
            s.push(c);
        }else{
            if(s.empty()){
                return false;
            }
            s.pop();
        }
    }
    return true;
}

string reorganize_u(string u){
    string reorg_u = "";
    for(int i = 1; i < u.length()-1; i++){
        char c = u[i];
        if(c == '('){
            reorg_u += ')';
        }else{
            reorg_u += '(';
        }
    }
    return reorg_u;
}

string reorganize_p(string p){
    if(p == ""){
        return p;
    }
    int l_count = 0;
    int r_count = 0;
    int s_point = p.length();
    for(int i = 0; i < p.size(); i++){
        char c = p[i];
        if(c == '('){
            l_count++;
        }else{
            r_count++;
        }
        if(l_count == r_count){
            s_point = i+1;
            break;
        }
    }
    string u = p.substr(0, s_point);
    string v = s_point != p.length() ? p.substr(s_point, p.length()-1) : "";
    if(check_correct_str(u)){
        return u + reorganize_p(v);
    }else{
        return "(" + reorganize_p(v) + ")" + reorganize_u(u);
    }
}

string solution(string p) {
    return reorganize_p(p);
}

int main(void){
    cout << solution("()))((()") << "\n";
}